modules/gdscript/editor/gdscript_docgen.cpp

```diff
@@ -36,40 +36,91 @@ using GDP = GDScriptParser;
+		if (!curr_class->identifier) { // All inner classes have an identifier, so this is the outer class.
 			return vformat("%s.%s", _get_script_path(curr_class->fqcn), full_name);
 		}
 		full_name = vformat("%s.%s", curr_class->identifier->name, full_name);
 	}
 	return full_name;
 }
 
-static PropertyInfo _property_info_from_datatype(const GDType &p_type) {
-	PropertyInfo pi;
-	pi.type = p_type.builtin_type;
-	if (p_type.kind == GDType::CLASS) {
-		pi.class_name = _get_class_name(*p_type.class_type);
-	} else if (p_type.kind == GDType::ENUM && p_type.enum_type != StringName()) {
-		pi.type = Variant::INT; // Only int types are recognized as enums by the EditorHelp
-		pi.usage |= PROPERTY_USAGE_CLASS_IS_ENUM;
-		// Replace :: from enum's use of fully qualified class names with regular .
-		pi.class_name = String(p_type.native_type).replace("::", ".");
-	} else if (p_type.kind == GDType::NATIVE) {
-		pi.class_name = p_type.native_type;
+static void _doctype_from_gdtype(const GDType &p_gdtype, String &r_type, String &r_enum, bool p_is_return = false) {
+	if (!p_gdtype.is_hard_type()) {
+		r_type = "Variant";
+		return;
+	}
+	switch (p_gdtype.kind) {
+		case GDType::BUILTIN:
+			if (p_gdtype.builtin_type == Variant::NIL) {
+				r_type = p_is_return ? "void" : "null";
+				return;
+			}
+			if (p_gdtype.builtin_type == Variant::ARRAY && p_gdtype.has_container_element_type()) {
+				_doctype_from_gdtype(p_gdtype.get_container_element_type(), r_type, r_enum);
+				if (!r_enum.is_empty()) {
+					r_type = "int[]";
+					r_enum += "[]";
+					return;
+				}
+				if (!r_type.is_empty() && r_type != "Variant") {
+					r_type += "[]";
+					return;
+				}
+			}
+			r_type = Variant::get_type_name(p_gdtype.builtin_type);
+			return;
+		case GDType::NATIVE:
+			r_type = p_gdtype.native_type;
+			return;
+		case GDType::SCRIPT:
+			if (p_gdtype.script_type.is_valid()) {
+				if (p_gdtype.script_type->get_global_name() != StringName()) {
+					r_type = _get_script_path(p_gdtype.script_type->get_global_name());
+					return;
+				}
+				if (!p_gdtype.script_type->get_path().is_empty()) {
+					r_type = _get_script_path(p_gdtype.script_type->get_path());
+					return;
+				}
+			}
+			if (!p_gdtype.script_path.is_empty()) {
+				r_type = _get_script_path(p_gdtype.script_path);
+				return;
+			}
+			r_type = "Object";
+			return;
+		case GDType::CLASS:
+			r_type = _get_class_name(*p_gdtype.class_type);
+			return;
+		case GDType::ENUM:
+			r_type = "int";
+			r_enum = String(p_gdtype.native_type).replace("::", ".");
+			if (r_enum.begins_with("res://")) {
+				r_enum = r_enum.trim_prefix("res://");
+				int dot_pos = r_enum.rfind(".");
+				if (dot_pos >= 0) {
+					r_enum = r_enum.left(dot_pos).quote() + r_enum.substr(dot_pos);
+				}
+			}
+			return;
+		case GDType::VARIANT:
+		case GDType::RESOLVING:
+		case GDType::UNRESOLVED:
+			r_type = "Variant";
+			return;
```

Comment:
```suggestion
		case GDType::VARIANT:
			r_type = "Variant";
			return;
		case GDType::RESOLVING:
			case GDType::UNRESOLVED:
			r_type = "<compilation-error>"
			return;
```

Should we consider something like this? This would make it so scripts that have errors would reflect that in the documentation. I think this could be useful in the case of, e.g. plugins that have errors because they need to be updated. It's not the user's code, and the errors are likely to only show up on editor open/project start, so they are harder to track down. At least this way users can notice the errors while looking at their plugin's documentation?

Reply:
Sounds reasonable, but I don't think it will work because of the hard type check at the beginning of the function. In addition, we do not have an agreement on what should be returned in case of type resolution error, we often return `VARIANT` dummy type.

Note: `DataType::to_string()` uses `<unresolved type>`.

https://github.com/godotengine/godot/blob/5444afae63f37cb4b890c781e5679a3b7ee5a392/modules/gdscript/gdscript_parser.cpp#L4324-L4326

Reply:
We would probably also have to change `EditorHelp` around so it would support compilation error formatting, so we can leave it as is, since it's the way it's always been, and consider this for a future PR?

Reply:
I'm not sure if we should be worried about this, it's unlikely that users will check compilation errors with the documentation rather than with the Debugger/Errors panel.